### DISCUSSION TITLE
Ability to define which home assistant user invoked a service call

### DISCUSSION THREAD

**Discussion by FeikoJoosten** [zachowj/node-red-contrib-home-assistant-websocket#1158] [Feature Request]
Hi there,

I was wondering if it could be possible to define which home assistant user invoked a service call. This would allow me to figure out which user toggled one of my devices from an ios action. This way it would also show up in the home assistant UI. 
![image](https://github.com/zachowj/node-red-contrib-home-assistant-websocket/assets/10920052/859e4385-46a4-428c-a527-ee7edbd8631d)
Top log is done by me toggling the device using the home assistant UI. 
Bottom log is done by me toggling the device using an ios action. Which in term invokes a service call using node-red.

One possible solution to this would be to allow dynamically overriding the server property of a node using for example `msg.payload.server`. Then create a server configuration for each home assistant with their own access token.

**Comment by zachowj** [comment]
Each server config opens a new connection to Home Assistant doubling the events that Node-RED receives and processes.

A better route might be to create a sub-flow that uses the REST API instead of the WebSocket one to make the service call with separate access tokens.

![image](https://github.com/zachowj/node-red-contrib-home-assistant-websocket/assets/37859597/0cdfa19e-2785-4ebc-a13c-690f8a29bdb9)

```
[{"id":"b0a633ceba2a2524","type":"subflow","name":"individual service call","info":"","category":"","in":[{"x":172,"y":208,"wires":[{"id":"c708531746dfda82"}]}],"out":[{"x":828,"y":192,"wires":[{"id":"401d6b9de1834f22","port":0}]}],"env":[],"meta":{},"color":"#DDAA99"},{"id":"c708531746dfda82","type":"switch","z":"b0a633ceba2a2524","name":"","property":"user","propertyType":"msg","rules":[{"t":"eq","v":"person1","vt":"str"},{"t":"eq","v":"person2","vt":"str"}],"checkall":"true","repair":false,"outputs":2,"x":290,"y":208,"wires":[["9c6a84667db86df3"],["2185d97436315a9d"]]},{"id":"401d6b9de1834f22","type":"ha-api","z":"b0a633ceba2a2524","name":"","server":"","version":1,"debugenabled":false,"protocol":"http","method":"post","path":"/api/services/{{payload.domain}}/{{payload.service}}","data":"payload.data","dataType":"jsonata","responseType":"json","outputProperties":[],"x":706,"y":192,"wires":[[]]},{"id":"9c6a84667db86df3","type":"change","z":"b0a633ceba2a2524","name":"access token for person 1","rules":[{"t":"set","p":"accessToken","pt":"msg","to":"AAAA","tot":"str"}],"action":"","property":"","from":"","to":"","reg":false,"x":488,"y":192,"wires":[["401d6b9de1834f22"]]},{"id":"2185d97436315a9d","type":"change","z":"b0a633ceba2a2524","name":"access token for person 2","rules":[{"t":"set","p":"accessToken","pt":"msg","to":"BBB","tot":"str"}],"action":"","property":"","from":"","to":"","reg":false,"x":488,"y":240,"wires":[["401d6b9de1834f22"]]},{"id":"6ceb54c7034b8860","type":"inject","z":"120358abd7c22d30","name":"person 1 turn on light","props":[{"p":"payload"},{"p":"user","v":"person1","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"{\"domain\": \"light\", \"service\": \"turn_on\", \"data\": { \"entity_id\": \"light.bed\"}}","payloadType":"json","x":732,"y":752,"wires":[["4c0bae97d0e1127a"]]},{"id":"4c0bae97d0e1127a","type":"subflow:b0a633ceba2a2524","z":"120358abd7c22d30","name":"","x":972,"y":752,"wires":[[]]},{"id":"19a6dfcff7299356","type":"inject","z":"120358abd7c22d30","name":"person 2 turn on light","props":[{"p":"payload"},{"p":"user","v":"person2","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"{\"domain\": \"light\", \"service\": \"turn_on\", \"data\": { \"entity_id\": \"light.bed\"}}","payloadType":"json","x":732,"y":800,"wires":[["4c0bae97d0e1127a"]]}]
```

**Reply by zachowj** [reply]
I messed that up. You would want to use a HTTP request node in place of the HA API node.

**Reply by FeikoJoosten** [reply]
Ah nice catch about the double events. The downside of the http request approach is that you can't get the fancy node dropdowns. Which in itself isn't that big a deal, but ideally a best of both worlds would be perfect. 
Edit: Which is also why I raised this discussion initially as I had already realized it should be possible using the http request approach. (Should've clarified that).
